File: CodingInterview/Programmers/Cpp/LV1/77_소수_찾기.cpp
// https://school.programmers.co.kr/learn/courses/30/lessons/12921?language=cpp
// 출처: 연습문제
// 대응 Python 풀이: ../../Python/LV1/!77_소수_찾기.py

// Number of times solving questions: 1
// 2026-09-16
// 소요 시간: -

#include "../test_runner.h"
// 아래는 프로그래머스 제출 코드와 동일하게 유지한다.
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(int n) {
    int answer = 0;

    vector<bool> prime(n+1, true);
    prime[0] = false;
    prime[1] = false;

    for (int i=2; i*i<=n; i++) {
        if (prime[i]) {
            for (int j=i*i; j<=n; j+=i) {
                prime[j] = false;
            }
        }
    }

    answer = count(prime.begin(), prime.end(), true);
    return answer;
}

int main() {
    Tester t("solution");
    t.check(solution(10), 4);
    t.check(solution(5), 3);

    // 아래는 직접 추가한 경계 케이스
    t.check("최소 입력 n=2", solution(2), 1);
    t.check("n=4 (완전제곱수, 2*2 가 지워져야 한다)", solution(4), 2);
    t.check("n=25 (5*5 가 지워져야 한다)", solution(25), 9);
    t.check("최대 입력 n=1,000,000", solution(1000000), 78498);
    return t.report();
}

File: CodingInterview/Programmers/Python/LV1/77_소수_찾기.py
# https://school.programmers.co.kr/learn/courses/30/lessons/12921?language=python3
# 출처: 연습문제

# Number of times solving questions: 1
# 2026-09-23

import sys, pathlib
sys.path.append(str(pathlib.Path(__file__).resolve().parent.parent))  # test_runner 경로
from test_runner import run_tests

import math

def is_prime(number):
    for i in range(2, number):
        if number%i == 0:
            return False
    return True

def solution(n):
    visited = [False for _ in range(n+1)]
    visited[0] = True
    visited[1] = True
    sqrt_number = int(math.sqrt(n))

    for number in range(2, sqrt_number+1):
        if visited[number]:
            continue
        if is_prime(number):
            for i in range(number, n+1):
                if i*number>n:
                    break
                visited[i*number] = True
    return visited.count(False)

# ─────────────────────────────────────────────────────────────
# 테스트
# ─────────────────────────────────────────────────────────────
test_cases = [
    (10, 4),
    (5, 3),

    # 아래는 직접 추가한 경계 케이스
    (2, 1),                 # 최소 입력
    (3, 2),
    (4, 2),                 # 4 = 2*2 는 소수가 아니다
    (25, 9),                # 25 = 5*5 — sqrt 상한 처리를 확인한다
    (1000000, 78498),       # 제한 최대
]

run_tests(solution, test_cases)
